src/mo_frontend/typing.ml

```diff
@@ -1998,8 +2000,11 @@ and check_cases env t_pat t cases =
 
 and check_case env t_pat t case =
   let {pat; exp} = case.it in
+  let initial_usage = enter_scope env in
   let ve = check_pat env t_pat pat in
-  recover (check_exp (adjoin_vals env ve) t) exp
+  let t' = recover (check_exp (adjoin_vals env ve) t) exp in
+  leave_scope env ve initial_usage;
```

Comment:
See above - perhaps leave_scope should only happen under `recover (...)`.

Reply:
Or we could postpone the unused warnings after successful type checking. This would probably be a quite general solution.

Reply:
Solved by merged https://github.com/dfinity/motoko/pull/4561